File: benchmarks/solutions/swift/r4-concurrent-fetch.swift
import Foundation

let urls = [
    "https://httpbin.org/get",
    "https://httpbin.org/status/404",
    "https://httpbin.org/delay/1",
    "https://httpbin.org/bytes/1024",
    "https://invalid.example.test",
]

await withTaskGroup(of: String.self) { group in
    let semaphore = AsyncSemaphore(count: 4)
    for url in urls {
        group.addTask {
            await semaphore.wait()
            defer { semaphore.signal() }
            do {
                let (data, response) = try await URLSession.shared.data(from: URL(string: url)!)
                let status = (response as! HTTPURLResponse).statusCode
                return "\(url): \(status) (\(data.count) bytes)"
            } catch {
                return "\(url): error: \(error.localizedDescription)"
            }
        }
    }
    for await result in group {
        print(result)
    }
}

actor AsyncSemaphore {
    private var count: Int
    private var waiters: [CheckedContinuation<Void, Never>] = []

    init(count: Int) { self.count = count }

    func wait() async {
        if count > 0 {
            count -= 1
        } else {
            await withCheckedContinuation { waiters.append($0) }
        }
    }

    func signal() {
        if waiters.isEmpty {
            count += 1
        } else {
            waiters.removeFirst().resume()
        }
    }
}

File: benchmarks/solutions/rust/r4-concurrent-fetch.rs
use std::sync::Arc;
use tokio::sync::Semaphore;

#[tokio::main]
async fn main() {
    let urls = vec![
        "https://httpbin.org/get",
        "https://httpbin.org/status/404",
        "https://httpbin.org/delay/1",
        "https://httpbin.org/bytes/1024",
        "https://invalid.example.test",
    ];

    let sem = Arc::new(Semaphore::new(4));
    let mut handles = Vec::new();

    for url in urls {
        let sem = sem.clone();
        let handle = tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            match reqwest::get(url).await {
                Ok(resp) => {
                    let status = resp.status().as_u16();
                    let bytes = resp.bytes().await.map(|b| b.len()).unwrap_or(0);
                    println!("{url}: {status} ({bytes} bytes)");
                }
                Err(e) => println!("{url}: error: {e}"),
            }
        });
        handles.push(handle);
    }

    for h in handles {
        let _ = h.await;
    }
}
